### PR TITLE
Reviews - Controller and Tests finished. 

### PR DESCRIPTION
**Pull request by dnsanche**
# bEtsy
Congratulations! You're submitting your assignment! These comprehension questions should be answered by all members of your team, not by a single teammate.

## Comprehension Questions
Question | Answer
:------------- | :-------------
Each team member: what is one thing you were primarily responsible for that you're proud of? | 
Each team member: what is one thing you were primarily responsible for that you would like targeted feedback on? | 
How did your team break up the work to be done? | 
How did your team utilize git to collaborate? | 
What did your group do to try to keep your code DRY while many people collaborated on it? | 
What was a technical challenge that you faced as a group? | 
What was a team/personal challenge that you faced as a group? | 
What was your application's ERD? (include a link) | 
What is your Trello URL? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/controllers/reviews_controller.rb [hunk 1]
```diff
@@ -1,16 +1,65 @@
...
-  
+
   def create
     @review = Review.new(review_params)
+    
+    if @review.save
+      redirect_to product_path(id: @review.product_id)
+      return
+    else 
+      flash.now[:failure] = "Work failed to save"
```
Comment (raisahv) [inline app/controllers/reviews_controller.rb]:
```suggestion
      flash.now[:failure] = "Review failed to save"
```

File: app/controllers/reviews_controller.rb [hunk 2]
```diff
@@ -1,16 +1,65 @@
...
   end
+
+  def find_review
+    @review = Review.find_by(id: params[:id])
+  end
+
+  def if_missing_review
+    if @review.nil?
+      flash[:error] = "Review with id #{params[:id]} was not found"
+      redirect_to root_path
```
Comment (NatalieTapias) [inline app/controllers/reviews_controller.rb]: do we want to redirect_to root_path here or render new?

File: app/controllers/reviews_controller.rb [hunk 3]
```diff
@@ -1,16 +1,65 @@
...
+
+  def edit ; end
+
+  def update    
+    if @review.update(review_params)
+      redirect_to product_path(id: @review.product_id)
+      return
+    else 
+      render :edit, status: :bad_request 
+      flash.now[:failure] = "Review failed to save"
```
Comment (raisahv) [inline app/controllers/reviews_controller.rb:35]:
```suggestion
      flash.now[:failure] = "Review failed to save"
      render :edit, status: :bad_request 
```

Comment (raisahv) [reply app/controllers/reviews_controller.rb:35]: Does the order of this matter? Do we need to flash before we render :edit?

File: app/controllers/reviews_controller.rb (hunk 2 above)
Comment (dtingg) [inline app/controllers/reviews_controller.rb]:
```suggestion
      redirect_back(fallback_location: root_path)
```

File: test/controllers/reviews_controller_test.rb [hunk 4]
```diff
@@ -1,7 +1,168 @@
...
+    end
+    
+    it "should respond with redirect with an invalid review" do 
+
+      invalid_id = -1
+      
+      get product_review_path(product_id: new_product.id, id: invalid_id)
+    
+      must_respond_with :redirect
+      must_redirect_to root_path
```
Comment (dtingg) [inline test/controllers/reviews_controller_test.rb]:
```suggestion
```

File: test/controllers/reviews_controller_test.rb [hunk 5]
```diff
@@ -1,7 +1,168 @@
...
+      must_respond_with :success
+    end
+    
+    it "should respond with redirect with an invalid review" do 
+      invalid_id = -1
+      
+      get product_review_path(product_id: new_product.id, id: invalid_id)
+    
+      must_respond_with :redirect
+      must_redirect_to root_path
```
Comment (dtingg) [inline test/controllers/reviews_controller_test.rb]:
```suggestion
```

File: test/controllers/reviews_controller_test.rb [hunk 6]
```diff
@@ -1,7 +1,168 @@
...
+      must_respond_with :success
+    end
+
+    it "will redirect if given invalid review ID" do
+      invalid_id = -1
+
+      get  edit_product_review_path(product_id: new_product.id, id: invalid_id)
+
+      must_respond_with :redirect
+      must_redirect_to root_path
```
Comment (dtingg) [inline test/controllers/reviews_controller_test.rb]:
```suggestion
```

**Review by NatalieTapias — COMMENTED** [review]
LGTM

File: app/controllers/reviews_controller.rb [hunk 7]
```diff
@@ -1,16 +1,66 @@
...
+  def show ; end
+
+  def edit ; end
+
+  def update    
+    if @review.update(review_params)
+      redirect_to product_path(id: @review.product_id)
+      return
+    else 
+      render :edit, status: :bad_request
```
Comment (raisahv) [inline app/controllers/reviews_controller.rb]:
```suggestion
```

File: test/controllers/reviews_controller_test.rb [hunk 8]
```diff
@@ -1,7 +1,165 @@
...
+
+    it "will not destroy a review if invalid id is provided" do
+      invalid_id = -1
+
+      expect {
+        delete product_review_path(product_id: new_product.id, id: invalid_id = -1)
+      }.wont_change "Review.count"
+
+      must_respond_with :redirect
+      must_redirect_to root_path
```
Comment (dtingg) [inline test/controllers/reviews_controller_test.rb]:
```suggestion
```

File: test/fixtures/reviews.yml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+review_one:
+  product_id: 123
```
Comment (raisahv) [inline test/fixtures/reviews.yml]:
```suggestion
  product: potter
```

File: test/fixtures/reviews.yml [hunk 10]
```diff
@@ -0,0 +1,21 @@
+review_one:
+  product_id: 123 
+  reviewer: "Anonymous Adie"
+  date: Time.now
+  rating: 4
+  comment: "OMG! I love this product.  Rated a four since the soap doesn't foam as much as I would like."
+review_two:
+  product_id: 456
```
Comment (raisahv) [inline test/fixtures/reviews.yml]:
```suggestion
  product: rose
```

File: test/fixtures/reviews.yml [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  reviewer: "Cynthia Tee"
+  date: Time.now
+  rating: 5
+  comment: "My favorite soap of all time!  I always give as part of birthday gifts!"
+review_three:
+  product_id: 789
```
Comment (raisahv) [inline test/fixtures/reviews.yml]:
```suggestion
  product: rose
```

**Review by NatalieTapias — COMMENTED** [review]
LGTM

**Review by dtingg — APPROVED** [review]

**Pull request merged**